core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,198 @@
+  private TableMetadata base;
+
+  UpdateSnapshotReferencesOperation(TableOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+    this.updatedRefs = Maps.newHashMap(base.refs());
+  }
+
+  @Override
+  public Map<String, SnapshotRef> apply() {
+    return refDiff(base.refs(), updatedRefs);
+  }
+
+  @Override
+  public void commit() {
+    TableMetadata updated = internalApply();
+    ops.commit(base, updated);
+  }
+
+  public UpdateSnapshotReferencesOperation createBranch(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    ValidationException.check(updatedRefs.get(name) == null, "Reference with name %s already exists", name);
+    SnapshotRef branch = SnapshotRef.branchBuilder(snapshotId).build();
+    SnapshotRef existingRef = updatedRefs.put(name, branch);
+    ValidationException.check(existingRef == null, "Reference with name %s already exists", name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation createTag(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Tag name cannot be null");
+    ValidationException.check(updatedRefs.get(name) == null, "Reference with name %s already exists", name);
+    SnapshotRef tag = SnapshotRef.tagBuilder(snapshotId).build();
+    SnapshotRef existingRef = updatedRefs.put(name, tag);
+    ValidationException.check(existingRef == null, "Reference with name %s already exists", name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation removeBranch(String name) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such branch with name %s", name);
+    ValidationException.check(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    updatedRefs.remove(name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation removeTag(String name) {
+    Preconditions.checkNotNull(name, "Tag name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such tag with name %s", name);
+    ValidationException.check(ref.isTag(), "Ref with name %s is a branch not a tag", name);
+    updatedRefs.remove(name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation renameBranch(String name, String newName) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    Preconditions.checkNotNull(newName, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    ValidationException.check(ref != null, "No such ref with name %s", name);
+    ValidationException.check(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    updatedRefs.put(newName, ref);
+    updatedRefs.remove(name, ref);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation replaceBranch(String name, long snapshotId) {
```

Comment:
Do you think we should also have a `replaceBranch(String target, String source)` that is like `rename`, but will replace the target branch?

Reply:
I think this makes sense but I was thinking we keep the API simpler initially, and when that's needed we can add it in a subsequent PR.  I'm not opinionated on this, we can also just add it here. 

Thoughts?

Reply:
I think replacing a branch with another is going to be common, so we may want to add a method for it. For example, people we talk to are planning to using branching and tagging to make changes, validate, and then commit them to main. That basically requires:
1. Create a branch from main
2. Commit to the branch
3. Validate the branch
4. Replace main with branch

So I'd probably have a `replaceBranch(String, String)` operation. And I would also have a `fastForwardBranch(String, String)` that validates replacing the branch will not lose any commits.

Reply:
That's a good point. In my mind I was thinking they would leverage cherry-picking . But actually, replacing a branch after validating is an easier semantic to understand. I will add that